### PR TITLE
Feature/#86 implement registries

### LINKED ISSUE #86: Implement the "Registries" module.
### Description

A module for resolving, registering, and updating did records.
The exposed methods are expected to look as follows:
```javascript

const registry = JolocomLib.registries.jolocom.create(...)
// Further configuration on the instance is possible.

const identityWallet = await registry.create(identityKey, ethreumKey)
const externalIdentity = await registry.resolve(did)
await registry.commit(wallet, ethereumKey)
const identityWallet = await registry.authenticate(identiyKey)
``` 

### TODO
- [x] Implement the defined module

### PR DESCRIPTION
**Pull request by nataschaberg**
Description

Fixes #86 .

> What does it solve?

> Were there any particular challenges that you faced and/or things that were not straightforward?

Checklist

- [ ] Keep flags added to the PR up to date
- [ ] Make sure the title describes the code well, description should contain Fixes #number_of_an_issue and a brief decscription of changes
- [ ] Make sure the PR is readable and contains changes only related to a single issue. Otherwise - divide into separate PRs
- [ ] Add new tests and make sure all tests are passing
- [ ] Make sure all edge cases are taken into account and covered by tests
- [ ] Run linter and fix all the errors and warnings

### CODE REVIEW

File: ts/registries/index.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import { JolocomRegistry } from './jolocomRegistry'
+
+class Registries {
+
```
Comment (charleenfei) [inline ts/registries/index.ts]: extra space?

**Comment by katszwn** [comment]
@charleenfei fixed :)

File: ts/identityWallet/identityWallet.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+export class IdentityWallet implements IIdentityWallet {
+  private didDocument: DidDocument
+  private privateIdentityKey: Buffer
+
+  public getDidDocument(): DidDocument {
+    return this.didDocument
+  }
+
+  public setDidDocument({didDocument}: {didDocument: DidDocument}): void {
```
Comment (Exulansis) [inline ts/identityWallet/identityWallet.ts]: Same point, wrapped argument

File: ts/identityWallet/identityWallet.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    this.didDocument = didDocument
+  }
+
+  public getPrivateIdentityKey(): Buffer {
+    return this.privateIdentityKey
+  }
+
+  public setPrivateIdentityKey({privateIdentityKey: Buffer}): void {
```
Comment (Exulansis) [inline ts/identityWallet/identityWallet.ts]: Same point

File: ts/identityWallet/types.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import { DidDocument } from '../identity/didDocument';
+
+export interface IIdentityWallet {
+  getDidDocument: () => DidDocument
+  setDidDocument: ({didDocument}: {didDocument: DidDocument}) => void
```
Comment (Exulansis) [inline ts/identityWallet/types.ts]: Same point

File: ts/identityWallet/types.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+  getPrivateIdentityKey: () => Buffer
+  setPrivateIdentityKey: ({privateIdentityKey}: {privateIdentityKey: Buffer}) => void
```
Comment (Exulansis) [inline ts/identityWallet/types.ts]: Same

File: ts/ipfs/index.ts [hunk 6]
```diff
@@ -5,11 +5,11 @@ import { IIpfsConnector, IIpfsConfig } from './types'
 export class IpfsStorageAgent implements IIpfsConnector {
```
Comment (Exulansis) [inline ts/ipfs/index.ts]: Same point with wrapped arguments.

File: ts/ipfs/types.ts [hunk 7]
```diff
@@ -5,8 +5,7 @@ export interface IIpfsConfig {
 }
 
 export interface IIpfsConnector {
```
Comment (Exulansis) [inline ts/ipfs/types.ts]: wrapped arguments

File: ts/registries/jolocomRegistry.ts [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+import { privateKeyToPublicKey, privateKeyToDID } from '../utils/crypto'
+import { DidDocument } from '../identity/didDocument'
+import { IDidDocumentAttrs } from '../identity/didDocument/types';
+
+/** Jolocom specific Registry, which uses IPFS
+ *  and Ethereum for registering the indentity and the resolution
+ *  mechanism.
+ */
+export class JolocomRegistry {
+  public ipfsConnector: IIpfsConnector
```
Comment (Exulansis) [inline ts/registries/jolocomRegistry.ts]: Do these need to be public?

File: ts/registries/jolocomRegistry.ts [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+  public ethereumConnector: IEthereumConnector
+
+  public static create(
```
Comment (Exulansis) [inline ts/registries/jolocomRegistry.ts]: Use interface, don't destruct, to have shorter, easier to read function signatures.

File: ts/registries/jolocomRegistry.ts [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    {ipfsConnector, ethereumConnector}:
+    {ipfsConnector: IIpfsConnector, ethereumConnector: IEthereumConnector}): JolocomRegistry {
+    const jolocomRegistry = new JolocomRegistry()
+    jolocomRegistry.ipfsConnector = ipfsConnector
+    jolocomRegistry.ethereumConnector = ethereumConnector
+
+    return jolocomRegistry
+  }
+
+  public async create(
```
Comment (Exulansis) [inline ts/registries/jolocomRegistry.ts]: Similar point as above.

File: ts/registries/jolocomRegistry.ts [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    const identityWallet = new IdentityWallet()
+    identityWallet.setDidDocument({didDocument: ddo})
+    identityWallet.setPrivateIdentityKey({privateIdentityKey})
+
+    await this.commit({wallet: identityWallet, privateEthereumKey})
+
+    return identityWallet
+  }
+
+  public async commit(
```
Comment (Exulansis) [inline ts/registries/jolocomRegistry.ts]: Similar point as above

File: ts/registries/jolocomRegistry.ts [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+      })
+    } catch (error) {
+      throw new Error(`Could not register DID record on Ethereum. ${error.message}`)
+    }
+  }
+
+  public async resolve({did}: {did: string}): Promise<IDidDocumentAttrs> {
+    // TODO: return an instance of Identity (which is extended DDO)
+    try {
+      return await this.ipfsConnector.catJSON({
```
Comment (Exulansis) [inline ts/registries/jolocomRegistry.ts]: Can be refactored to:
```javascript
const hash = await this.ethereumConnector.resolveDID(did)
return await this.ipfsConnector.catJSON(hash)
```

File: ts/registries/jolocomRegistry.ts [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+        hash: await this.ethereumConnector.resolveDID({did})
+      }) as Promise<IDidDocumentAttrs>
+    } catch (error) {
+      throw new Error(`Could not retrieve DID Document. ${error.message}`)
+    }
+  }
+
+  public async authenticate({privateIdentityKey}: {privateIdentityKey: Buffer}): Promise<IdentityWallet> {
```
Comment (Exulansis) [inline ts/registries/jolocomRegistry.ts]: Same issue with wrapped arguments.

File: ts/registries/jolocomRegistry.ts [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+    const identityWallet = new IdentityWallet()
+    const did = privateKeyToDID(privateIdentityKey)
+    // TODO: change according to return of resolve
+    const ddoAttrs = await this.resolve({did})
+    const didDocument = DidDocument.fromJSON(ddoAttrs)
+    identityWallet.setDidDocument({didDocument})
+    identityWallet.setPrivateIdentityKey({privateIdentityKey})
```
Comment (Exulansis) [inline ts/registries/jolocomRegistry.ts]: same

File: tests/jolocomRegistry.test.ts [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+    protocol: 'testprotocol'
+  }})
+
+  const ethereumConnector = new EthResolver({config: {
+    providerUrl: 'testEthereumURL',
+    contractAddress: '0x8389B5a24a1c56aFAD7309EF3b8e04bBadC935c4'
+  }})
+
+  const unixEpoch = moment.utc('2018-01-24T15:42:15Z', moment.ISO_8601).valueOf()
+  const clock = lolex.install({now: unixEpoch})
```
[the remark this answers is not in the document]
Comment (Exulansis) [reply tests/jolocomRegistry.test.ts]: Perhaps using `clock.uninstall()` in afterEach hooks.

[the remark this answers is not in the document]
Comment (Exulansis) [reply tests/jolocomRegistry.test.ts]: Also we could try use the fake timers provided with sandboxes, as opposed to lolex.

File: tests/data/identity.ts [hunk 16]
```diff
@@ -60,6 +60,24 @@ export default {
       "expires": "2018-10-24T15:42:15Z"
     },
   "signature": "26f1096e286f82955c7ca84fc9ecbbb2b9524577c735b08551134a4091df74563b1d2fdc1f5c304c6e9c9209794ff7cc78fe6bec1ceef1275220f8b4a0a81e6a"
+  },
```
Comment (Exulansis) [inline tests/data/identity.ts]: There are multiple linting issues in this document, mostly double quotes, long lines, inconsistent spacing

Comment (Exulansis) [reply tests/data/identity.ts]: Default exports should be avoided as well.

File: ts/identityWallet/identityWallet.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (katszwn) [reply ts/identityWallet/identityWallet.ts]: I think it makes sense for the sake of readability

File: ts/ethereum/index.ts [hunk 17]
```diff
@@ -1,17 +1,20 @@
...
+    this.ethResolver = new EthereumResolver(config.contractAddress, config.providerUrl)
   }
 
-  public async resolveDID(did: string): Promise<string> {
+  public async resolveDID({did}: {did: string}): Promise<string> {
     return this.ethResolver.resolveDID(did)
   }
 
-  public async updateDIDRecord(ethereumKey: object, did: string, newHash: string): Promise<void> {
+  public async updateDIDRecord(
```
[the remark this answers is not in the document]
Comment (katszwn) [reply ts/ethereum/index.ts]: Prettier would fix those! :D

File: tests/data/identity.ts (hunk 16 above)
Comment (katszwn) [reply tests/data/identity.ts]: Yep, but that's not the scope of this PR - can we just run Prettier on everything? I can open an issue if you think that makes sense :)

**Comment by Exulansis** [comment]
The tests still need some extra work, but we can leave this for later, in the interest of time.
